sdcflows/transform.py

```diff
@@ -72,151 +274,271 @@ def fit(self, spatialimage):
 
         """
         # Calculate the physical coordinates of target grid
-        if isinstance(spatialimage, (str, bytes, Path)):
-            spatialimage = nb.load(spatialimage)
+        if isinstance(target_reference, (str, bytes, Path)):
+            target_reference = nb.load(target_reference)
+
+        approx &= xfm_data2fmap is not None  # Approximate iff xfm_data2fmap is defined
+        xfm_data2fmap = xfm_data2fmap if xfm_data2fmap is not None else np.eye(4)
+        target_affine = target_reference.affine.copy()
+        target_header = target_reference.header.copy()
+
+        if approx:
+            _tmp_shape = target_reference.shape[:3]
+
+        # Project the reference's grid onto the fieldmap's
+        target_reference = target_reference.__class__(
+            target_reference.dataobj,
+            xfm_data2fmap @ target_affine,
+            target_header,
+        )
 
-        # Initialize xform (or set identity)
-        xfm = self.xfm if self.xfm is not None else nt.linear.Affine()
+        # Make sure the data array has all cosines positive (i.e., no axes are flipped)
+        target_reference, _ = ensure_positive_cosines(target_reference)
+
+        # Approximate only iff the coordinate systems are not aligned
+        finest_coeffs = listify(self.coeffs)[-1]
+        approx &= not np.allclose(
+            np.linalg.norm(
+                np.cross(finest_coeffs.affine[:-1, :-1].T, target_reference.affine[:-1, :-1].T),
+                axis=1,
+            ),
+            0,
+            atol=1e-3,
+        )
 
-        if self.mapped is not None:
-            newaff = spatialimage.affine
-            newshape = spatialimage.shape
+        # TODO Separate cache validation from here. With the resampling, this
+        # code will always  determine the cache must be recalculated.
+        # if self.mapped is not None:
+        #     newshape = target_reference.shape
 
-            if np.all(newshape == self.mapped.shape) and np.allclose(
-                newaff, self.mapped.affine
-            ):
-                return False
+        #     if np.all(newshape == self.mapped.shape) and np.allclose(
+        #         target_affine, self.mapped.affine
+        #     ):
+        #         return False
 
         weights = []
         coeffs = []
+        _tmp_reference = None
+
+        hdr = target_reference.header.copy()
+        if approx:
+            from sdcflows.utils.tools import deoblique_and_zooms
+
+            # Generate a sampling reference on the fieldmap's space that fully covers
+            # the target_reference's grid.
+            target_reference = deoblique_and_zooms(
+                listify(self.coeffs)[-1],
+                target_reference,
+            )
 
         # Generate tensor-product B-Spline weights
         for level in listify(self.coeffs):
```

Comment:
`listify` is potentially called 3 times on `self.coeffs` - perhaps we save the cycles and just set to a variable?